File: cpp/rpc/src/rpc_client.cpp
#include "rpc_client.hpp"

#include <iostream>
#include <nlohmann/json.hpp>

CppHttpLibClientConnector::CppHttpLibClientConnector(const std::string &host, int port, double timeout) : _httpClient(host.c_str(), port), _timeout(timeout) {

}

std::string CppHttpLibClientConnector::Send(const std::string &request){
    this->_httpClient.set_timeout_sec(this->_timeout);
    auto res = this->_httpClient.Post("/", request, "application/json");
    if (!res || res->status != 200) {
        throw jsonrpccxx::JsonRpcException(-32003, "client connector error, received status != 200");
    }
    return res->body;
}

bool rpc_call(const std::string &address, unsigned port, const std::string &method, const jsonrpccxx::positional_parameter& request, nlohmann::json& response, long timeout, unsigned id){
    try{
        CppHttpLibClientConnector httpClient(address, port);
        jsonrpccxx::JsonRpcClient client(httpClient, jsonrpccxx::version::v2);
        response = client.CallMethod<nlohmann::json>(1,method,request);
        return true;
    }catch(const nlohmann::json::exception& e){
        std::cout<<e.what()<<std::endl;
        return false;
    }catch(const jsonrpccxx::JsonRpcException& e){
        std::cout<<e.what()<<std::endl;
        return false;
    }
}


int main(){

    nlohmann::json response;
    // Example for a call without arguments.
    rpc_call("localhost",8900,"hello_world",{},response);
    // Example for a call with two arguments of type double.
    rpc_call("localhost",8900,"add",{4.0,5.0},response);

    // Print the response.
    std::cout<<"Result is "<<response["result"]<<std::endl;
}

File: python/rpc/rpc_client.py
import requests
import json


def rpc_call(url, method, params=None, timeout=100):
    """
    This function calls a method at an RPC server at the given url.
    :param url: The url must be a string consisting of address:port. Address is a hostname or an IP.
    :param method: The method you want to call at the remote server. Must be a string.
    :param params: The parameters you want to provide to the given method. Must be a dict.
    :param timeout: The function waits this amount of time for an answer from the server before terminating.
    :return: Returns the response from the server, returns None in case of an error.
    """
    headers = {'content-type': 'application/json'}

    payload = {
        "id": 1,
        "jsonrpc": "2.0",
        "method": method,
        "params": params,
    }
    print(payload)
    try:
        response = requests.post(url, data=json.dumps(payload), headers=headers, timeout=timeout).json()
    except requests.Timeout:
        print('Timeout, server has terminated or does not exist.')
        response = None
    except requests.ConnectionError:
        print('Connection error, target url not reachable.')
        response = None

    return response


if __name__ == '__main__':
    print(rpc_call('http://localhost:8900', 'hello_world', None))  # An example call without any arguments.
    params={'a': 2,'b': 6}
    print(rpc_call('http://localhost:8900', 'add', params))  # An example call with two arguments.
